feat(publish): Added Github Actions output

jbnjohnathan opened the pull request:
- Added output for `released` and `version` for Github Actions
`released` is either `true` if the version was bumped, or `false` otherwise
`version` is the new version if version was bumped, otherwise the current version
- Re-ordered imports with `isort`
- Added tests

Still having some issues with some of the tests.  @sp-luciano-chinke  Could you help out?


Inspired by: https://github.com/python-semantic-release/python-semantic-release/issues/401

bernardcooke53 commented:
Hey both :wave:

This is looking promising! There's a couple of things that I think would be nice to look at as part of this PR:
* I think it would be great to declare those outputs in `action.yml` too - see [docs](https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#outputs-for-docker-container-and-javascript-actions). A quick comment explaining that the idea is to do `echo "<key>=<value>" | tee -a $GITHUB_OUTPUT` (but in Python) would make it a bit clearer in my opinion.
* Could consider adding to the [automatic-releases](https://github.com/python-semantic-release/python-semantic-release/blob/master/docs/automatic-releases/github-actions.rst) page a table about the outputs
* Maybe it would be useful to include `level_bump` as an output - although I think `change_type` is possibly a better name for someone not familiar with the internal details

I don't think adding the output to the `version` command will be useful at the moment since the action only runs `publish`, although some people do just use `version` as part of their CI.

Let me know what your thoughts are - and let's see what's up with those tests :wrench:

jbnjohnathan commented:
> Hey both 👋
> 
> This is looking promising! There's a couple of things that I think would be nice to look at as part of this PR:
> 
> * I think it would be great to declare those outputs in `action.yml` too - see [docs](https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#outputs-for-docker-container-and-javascript-actions). A quick comment explaining that the idea is to do `echo "<key>=<value>" | tee -a $GITHUB_OUTPUT` (but in Python) would make it a bit clearer in my opinion.
> * Could consider adding to the [automatic-releases](https://github.com/python-semantic-release/python-semantic-release/blob/master/docs/automatic-releases/github-actions.rst) page a table about the outputs
> * Maybe it would be useful to include `level_bump` as an output - although I think `change_type` is possibly a better name for someone not familiar with the internal details
> 
> I don't think adding the output to the `version` command will be useful at the moment since the action only runs `publish`, although some people do just use `version` as part of their CI.
> 
> Let me know what your thoughts are - and let's see what's up with those tests 🔧

Sure thing : ) I'll check all of that out and update the PR

jbnjohnathan commented:
> Hey both 👋
> 
> This is looking promising! There's a couple of things that I think would be nice to look at as part of this PR:
> 
> * I think it would be great to declare those outputs in `action.yml` too - see [docs](https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#outputs-for-docker-container-and-javascript-actions). A quick comment explaining that the idea is to do `echo "<key>=<value>" | tee -a $GITHUB_OUTPUT` (but in Python) would make it a bit clearer in my opinion.
> * Could consider adding to the [automatic-releases](https://github.com/python-semantic-release/python-semantic-release/blob/master/docs/automatic-releases/github-actions.rst) page a table about the outputs
> * Maybe it would be useful to include `level_bump` as an output - although I think `change_type` is possibly a better name for someone not familiar with the internal details
> 
> I don't think adding the output to the `version` command will be useful at the moment since the action only runs `publish`, although some people do just use `version` as part of their CI.
> 
> Let me know what your thoughts are - and let's see what's up with those tests 🔧

I added the outputs to the `action.yml` file now (I was very sure I had already but must have missed it)
Also added to the `githut-actions.srt` doc

I didn't quite follow what the `level_bump` output was going to be for, so I think this PR will only be for the `released` and `version` outputs.

Also I didn't get the `I don't think adding the output to the version command will be useful at the moment since the action only runs publish,` comment. Could you clarify?

jbnjohnathan commented:
I moved my code into the function `should_bump_version` instead, I think it was a better fit.

bernardcooke53 commented on semantic_release/cli.py:
```diff
@@ -220,13 +220,21 @@ def should_bump_version(
     """Test whether the version should be bumped."""
     if new_version == match_version and not retry:
         logger.info("No release will be made.")
+        if os.getenv("GITHUB_ACTIONS"):
```
This is repeated a few times - I personally disagree with "hidden" parameters (e.g. loading env vars) in a function, but I would consider writing a little helper like 
```python
def write_github_output(output: io.StringIO, released: bool, version: str) -> None:
    print(f'released={str(released).lower()}', file=output)
    print(f'version={match_version}', file=output)
```
You can also use a little trick (thanks, Python!) to condense the output logic if you want:

```python
with open(os.getenv("GITHUB_OUTPUT", os.devnull), "a") as f:
    ...
```
That will just send the output to the null file if "GITHUB_OUTPUT" isn't set; it strictly shouldn't be necessary to even check "GITHUB_ACTIONS".

The advantages are that you can open the file once and write to it multiple times, if need be, and that when testing you can directly pass a `io.StringIO` instance and check the content after calling this function - no files needed.

That said, I don't mind terribly if you'd prefer to just pass the filename as a string and open the file within the function - both should be simple enough to test (see comment on tests)

bernardcooke53 commented on tests/test_cli.py:
```diff
@@ -1,19 +1,40 @@
 def runner():
     return CliRunner()
 
+# Only mock open for github_output_file
```
I'm not sure I'm keen on this - I think a better strategy is to use `pytest`'s [`tmp_path`](https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html) fixture to set up a temporary file fixture, set `GITHUB_OUTPUT` equal to the absolute path of that file, and then check the contents of the file after `publish`.

That way if we decide to refactor internally, for example using `pathlib` instead, the tests still pass - introducing special logic in this mock function doesn't feel right to me.

This might be different again if you chose to use that utility function taking a `StringIO` buffer directly in my other comment - I think both would result in tests that are slightly more robust though.

bernardcooke53 commented on tests/test_cli.py:
```diff
@@ -1,19 +1,40 @@
-from click.testing import CliRunner
+from unittest.mock import ANY, mock_open
```
`pytest` is generally the framework of choice for our tests - only the really really old ones still use `unittest` if we can avoid leaning on it further that would be ideal :pray:

bernardcooke53 commented on docs/automatic-releases/github-actions.rst:
```diff
@@ -23,6 +23,16 @@ Inputs
 | ``additional_options``   | Additional :ref:`cmd-common-options` for the ``publish`` command. Example: ``--noop``  |
 +--------------------------+----------------------------------------------------------------------------------------+
 
++--------------------------+----------------------------------------------------------------------------------------+
```
:+1: thank you!

bernardcooke53 commented on action.yml:
```diff
@@ -36,6 +36,12 @@ inputs:
     description: 'The ssh private key used to sign commits'
     required: false
 
+outputs:
```
:+1: thanks for adding this!

bernardcooke53 commented on docs/automatic-releases/github-actions.rst:
```diff
@@ -23,6 +23,16 @@ Inputs
 +--------------------------+----------------------------------------------------------------------------------------+
 
++--------------------------+----------------------------------------------------------------------------------------+
+| Outpput                  | Description                                                                            |
```
typo: ~`Outpput`~ `Output`

bernardcooke53 commented on semantic_release/cli.py:
I am not sure I agree that the code belongs in `should_bump_version` - that function should really only identify if a new version should be created, and I don't like introducing a side effect into it. I know that that's the case with several other functions in the codebase but I don't want to make the problem any worse - the suggestion about creating a utility should cut down on the amount of code needed at the call site, but ideally I'd like this in the top-level CLI commands please :pray:

bernardcooke53 reviewed (commented):
Apologies for taking a little while to come back on this. I have a few suggestions for neatening the tests and I think the logic needs to come out of `should_bump_version` :pray: thank you for the effort so far!

bernardcooke53 commented:
I have brought this into v8 now, so I think it's easiest to close this & raise a new PR if any of the outputs need changing - documented [here](https://python-semantic-release.readthedocs.io/en/latest/automatic-releases/github-actions.html#setting-up-python-semantic-release-on-github-actions).

Thank you @jbnjohnathan for raising this & pulling the code together for this PR - very much appreciated :1st_place_medal:
